### PR TITLE
KAFKA-7296; Handle coordinator loading error in TxnOffsetCommit

### PR DESCRIPTION
**Pull request by hachikuji**
We should check TxnOffsetCommit responses for the COORDINATOR_LOADING_IN_PROGRESS error code and retry if we see it. Additionally, if we encounter an abortable error, we need to ensure that pending transaction offset commits are cleared.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 1]
```diff
@@ -1272,50 +1272,49 @@ String coordinatorKey() {
...
             Map<TopicPartition, Errors> errors = txnOffsetCommitResponse.errors();
 
+            log.debug("Received TxnOffsetCommit response for consumer group {}: {}", builder.consumerGroupId(),
+                    errors);
+
             for (Map.Entry<TopicPartition, Errors> entry : errors.entrySet()) {
                 TopicPartition topicPartition = entry.getKey();
                 Errors error = entry.getValue();
                 if (error == Errors.NONE) {
-                    log.debug("Successfully added offsets {} from consumer group {} to transaction.",
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1282]: Just curious why removing this debug entry?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 2]
```diff
@@ -1272,50 +1272,49 @@ String coordinatorKey() {
...
+                    break;
                 }
             }
 
-            if (!hadFailure || !result.isSuccessful()) {
-                // all attempted partitions were either successful, or there was a fatal failure.
-                // either way, we are not retrying, so complete the request.
+            if (result.isCompleted()) {
+                pendingTxnOffsetCommits.clear();
+            } else if (pendingTxnOffsetCommits.isEmpty()) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1312]: Is this ever possible? `result` indicate retries needed while `pendingTxnOffsetCommits.isEmpty()`?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 3]
```diff
@@ -2399,7 +2413,9 @@ public boolean matches(AbstractRequest body) {
         };
     }
 
-    private void prepareAddOffsetsToTxnResponse(Errors error, final String consumerGroupId, final long producerId,
+    private void prepareAddOffsetsToTxnResponse(Errors error,
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: First parameter can be final as well?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 4]
```diff
@@ -2449,7 +2465,7 @@ private void doInitTransactions(long pid, short epoch) {
 
     private void assertAbortableError(Class<? extends RuntimeException> cause) {
         try {
-            transactionManager.beginTransaction();
+            transactionManager.beginCommit();
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:2467]: Is this relevant to the PR?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1282]: Two reasons:
1. It is misleading. There may be some partitions that have failed, but we still print all offsets that were in the request.
2. It is redundant. We are printing all offsets in the request for every successfully committed offset.

I replaced it with the log message above which just prints the errors in one shot.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1312]: This is where we complete the result if there are no errors.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java (hunk 4 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:2467]: Not directly, but this seemed like a better way to verify an abortable error. We expect `beginTransaction()` to fail regardless whether the error is abortable or fatal.

**Review by guozhangwang — APPROVED** [review]
LGTM!

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 5]
```diff
@@ -1272,50 +1272,49 @@ String coordinatorKey() {
         public void handleResponse(AbstractResponse response) {
             TxnOffsetCommitResponse txnOffsetCommitResponse = (TxnOffsetCommitResponse) response;
             boolean coordinatorReloaded = false;
-            boolean hadFailure = false;
             Map<TopicPartition, Errors> errors = txnOffsetCommitResponse.errors();
 
+            log.debug("Received TxnOffsetCommit response for consumer group {}: {}", builder.consumerGroupId(),
+                    errors);
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1278]: Do you think it would make sense to put an `if (!errors.isEmpty())` check around this log line for the cases when there are no errors? If I see it correctly it'd print `Received TxnOffsetCommit response for consumer group my-consumer-group:` which is I think is a bit misleading, and also I suppose it would be empty most of the times.

**Review by viktorsomogyi — APPROVED** [review]
Left one minor comment on the logging bit but otherwise it looks good.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1278]: The `errors()` method name is a little misleading. The map contains both the successful partitions as well as those that failed. Most of the time, we would just see `Errors.NONE`.

**Comment by hachikuji** [comment]
@viktorsomogyi Thanks for reviewing. I will go ahead and merge, but please follow up if my explanation does not make sense.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 5 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:1278]: Ok, I think I see now, thanks for the explanation.
